common/src/main/java/com/wynntils/features/map/WorldMarkersFeature.java

```diff
@@ -0,0 +1,513 @@
+
+    private final List<RenderedMapLocation> renderedMapLocations = new ArrayList<>();
+    private CustomColor currentRainbowColor = CommonColors.RED;
+
+    @SubscribeEvent
+    public void onTick(TickEvent event) {
+        int r = currentRainbowColor.r;
+        int g = currentRainbowColor.g;
+        int b = currentRainbowColor.b;
+
+        if (r > 0 && b == 0) {
+            r -= RAINBOW_CHANGE_RATE;
+            g += RAINBOW_CHANGE_RATE;
+        }
+        if (g > 0 && r == 0) {
+            g -= RAINBOW_CHANGE_RATE;
+            b += RAINBOW_CHANGE_RATE;
+        }
+        if (b > 0 && g == 0) {
+            r += RAINBOW_CHANGE_RATE;
+            b -= RAINBOW_CHANGE_RATE;
+        }
+
+        r = MathUtils.clamp(r, 0, 255);
+        g = MathUtils.clamp(g, 0, 255);
+        b = MathUtils.clamp(b, 0, 255);
+        currentRainbowColor = new CustomColor(r, g, b, 255);
+    }
+
+    // Rendered map location calculation happens here
+    // Beacon beam rendering happens here
+    @SubscribeEvent
+    public void onRenderLevelLast(RenderTileLevelLastEvent event) {
+        // We update the rendered map locations here so that we can render the beacon beams in the same event
+        // and also use it below to render the icons and labels
+        updateRenderedMapLocations(event.getProjectionMatrix(), event.getCamera());
+        if (renderedMapLocations.isEmpty()) return;
+
+        PoseStack poseStack = event.getPoseStack();
+
+        // render beacon beams
+        for (RenderedMapLocation mapLocationPair : renderedMapLocations) {
+            double alpha = mapLocationPair.visibility();
+            MapLocation mapLocation = mapLocationPair.mapLocation();
+
+            ResolvedMapAttributes resolvedMapAttributes = Services.MapData.resolveMapAttributes(mapLocation);
+            ResolvedMarkerOptions resolvedMarkerOptions = resolvedMapAttributes.markerOptions();
+
+            Position camera = event.getCamera().getPosition();
+            Location location = mapLocation.getLocation();
+
+            double dx = location.x - camera.x();
+            double dy = location.y - camera.y();
+            double dz = location.z - camera.z();
+
+            double distance = MathUtils.magnitude(dx, dz);
+            int maxDistance = McUtils.options().renderDistance().get() * 16;
+
+            if (distance > maxDistance) {
+                double scale = maxDistance / distance;
+
+                dx *= scale;
+                dz *= scale;
+            }
+
+            poseStack.pushPose();
+            poseStack.translate(dx, dy, dz);
+
+            CustomColor color = resolvedMarkerOptions.beaconColor();
+
+            if (color == CustomColor.NONE) {
+                // If the color is set to NONE, we skip rendering the beacon beam
+                // This is not logged as render time logging is heavy
+                continue;
+            }
+
+            int colorInt;
+            if (color == CommonColors.RAINBOW) {
```

Comment:
Can we express this in json btw?

Reply:
We can (and do) in built in providers. That's enough for me.